pkg/infrastructure/azure/azure.go

```diff
@@ -105,9 +112,117 @@ func (p *Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionI
+		subscriptionID,
+		tokenCredential,
+		&arm.ClientOptions{
+			ClientOptions: policy.ClientOptions{
+				Cloud: cloudConfiguration,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create armmsi client: %w", err)
+	}
+	_, err = armmsiClientFactory.NewUserAssignedIdentitiesClient().CreateOrUpdate(
+		ctx,
+		resourceGroupName,
+		userAssignedIdentityName,
+		armmsi.Identity{
+			Location: ptr.To(platform.Region),
+			Tags:     tags,
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create user assigned identity %s: %w", userAssignedIdentityName, err)
+	}
+	userAssignedIdentity, err := armmsiClientFactory.NewUserAssignedIdentitiesClient().Get(
+		ctx,
+		resourceGroupName,
+		userAssignedIdentityName,
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get user assigned identity %s: %w", userAssignedIdentityName, err)
+	}
+	principalID := *userAssignedIdentity.Properties.PrincipalID
+
+	logrus.Debugf("UserAssignedIdentity.ID=%s", *userAssignedIdentity.ID)
+	logrus.Debugf("PrinciapalID=%s", principalID)
+
+	clientFactory, err := armauthorization.NewClientFactory(
+		subscriptionID,
+		tokenCredential,
+		&arm.ClientOptions{
+			ClientOptions: policy.ClientOptions{
+				Cloud: cloudConfiguration,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create armauthorization client: %w", err)
+	}
+
+	roleDefinitionsClient := clientFactory.NewRoleDefinitionsClient()
+
+	var contributor *armauthorization.RoleDefinition
+	roleDefinitionsPager := roleDefinitionsClient.NewListPager(*resourceGroup.ID, nil)
+	for roleDefinitionsPager.More() {
+		roleDefinitionsList, err := roleDefinitionsPager.NextPage(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to find any role definitions: %w", err)
+		}
+		for _, roleDefinition := range roleDefinitionsList.Value {
+			if *roleDefinition.Properties.RoleName == "Contributor" {
+				contributor = roleDefinition
+				break
+			}
+		}
+	}
+	if contributor == nil {
+		return fmt.Errorf("failed to find contributor definition")
+	}
+
+	roleAssignmentsClient := clientFactory.NewRoleAssignmentsClient()
+	scope := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s", subscriptionID, resourceGroupName)
+	roleAssignmentUUID := uuid.New().String()
+
+	// XXX: Azure doesn't like creating an identity and immediately
+	// creating a role assignment for the identity. There can be
+	// replication delays. So, retry every 10 seconds for a minute until
+	// the role assignment gets created.
+	//
+	// See https://aka.ms/docs-principaltype
+	for i := 0; i < retryCount; i++ {
+		_, err = roleAssignmentsClient.Create(ctx, scope, roleAssignmentUUID,
+			armauthorization.RoleAssignmentCreateParameters{
+				Properties: &armauthorization.RoleAssignmentProperties{
+					PrincipalID:      ptr.To(principalID),
+					RoleDefinitionID: contributor.ID,
+				},
+			},
+			nil,
+		)
+		if err == nil {
+			break
+		}
+		time.Sleep(retryTime)
```

Comment:
Nit: if we move the sleep at the start of the for loop, we do the role assignment 10 secs after creating the identity. Likely to achieve the same results with fewer API calls.

Reply:
The tradeoff here is that it can possibly add more time. What do we care about more here? an extra API call, or additional 10 seconds ? I don't have a preference either way.

Reply:
It will be both. I don't have a strong opinion on this. Let us leave it as is.